runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -660,6 +660,107 @@ static void *dereferenceStructPointerChain(void *baseStruct, TR::Node *baseNode,
+         node->getFirstChild()->getFirstChild() :
+         node->getFirstChild();
+   // Abort if the indirection is more than a single level.
+   if (!addressChildNode->getOpCode().hasSymbolReference()
+       || addressChildNode != baseExpression)
+      return NULL;
+
+   // We only consider the case where isJavaField is true for verifyFieldAccess
+   if (isJavaField(symRef, comp))
+      {
+      TR_OpaqueClassBlock *fieldClass = NULL;
+
+      if (symRef->getCPIndex() < 0 &&
+          field->getRecognizedField() != TR::Symbol::UnknownField)
+         {
+         const char* className;
+         int32_t length;
+         className = field->owningClassNameCharsForRecognizedField(length);
+         fieldClass = fej9->getClassFromSignature(className, length, symRef->getOwningMethod(comp));
+         }
+      else
+         fieldClass = symRef->getOwningMethod(comp)->getDeclaringClassFromFieldOrStatic(comp,
+            symRef->getCPIndex());
+
+      TR_OpaqueClassBlock *objectClass =
+         fej9->getObjectClassFromKnownObjectIndex(comp, baseKnownObject);
+
+      // field access verified
+      if ((fieldClass != NULL) && (fej9->isInstanceOf(objectClass, fieldClass, true) == TR_yes))
+         {
+         // check the recognized fields case of avoidFoldingInstanceField
+         // the non-null checks are done when we get the actual values
+         if (field->getRecognizedField() == TR::Symbol::Java_lang_invoke_CallSite_target ||
+             field->getRecognizedField() == TR::Symbol::Java_lang_invoke_MethodHandle_form)
+             return NULL;
+
+         TR::DataType loadType = node->getDataType();
+
+         switch (loadType)
+            {
+            case TR::Int32:
+            case TR::Int64:
+            case TR::Float:
+            case TR::Double:
+               {
+               // not address
+               auto stream = comp->getStream();
+               stream->write(JITServer::MessageType::KnownObjectTable_getFieldAddressData,
+                             baseKnownObject, symRef->getOffset());
+               J9::TransformUtil::value value = std::get<0>(stream->read<J9::TransformUtil::value>());
+               *valuePtr = value;
+
+               return valuePtr;
+               }
+               break;
+            case TR::Address:
+               {
+               if (isFinalFieldPointingAtRepresentableNativeStruct(symRef, comp) ||
+                  isFinalFieldPointingAtNativeStruct(symRef, comp))
+                  {
+                  return NULL;
+                  }
+               else if (field->isCollectedReference())
+                  {
+                  auto stream = comp->getStream();
+                  stream->write(
+                     JITServer::MessageType::KnownObjectTable_addFieldAddressFromBaseIndex,
+                              baseKnownObject, symRef->getOffset());
+                  auto recv = stream->read<TR::KnownObjectTable::Index, uintptr_t *>();
```

Comment:
Why return an index instead of a value? I know that the JITServer is not supposed to receive object references, but if that reference is safe to be embedded in the native code, then it means that is not going to change.